### DISCUSSION TITLE
global Variable for sizing

### DISCUSSION THREAD

**Discussion by SReiter149** [francof2a/fxpmath#34] [Q&A] [answered]
Hi I'm new to Fxpmath,

I was digging around in the code and I noticed that the default sizing is optimal, I was wondering if there was an easy way to easily set all variables to output to 'same' sizing instead? I have too many variables to manually do that for everyone so I was hoping that you guys have a global variable that will let me set that as the default instead.

**Comment by SReiter149** [comment]
here is an example piece of code as to why I am asking this question:

`x = Fxp(np.random.random((1,400)),True,32,10)
y = Fxp(np.random.random((400,128)),True,32,10)
z = np.dot(x,y)
print(z.info(3))`

showing the info shows that the word bits are 73 and I understand that you can't have more than 64 bits because of NumPy but I was hoping there was a way to keep it below 64, preferably keeping it to 32 but I just want a result that isn't all 0s.

**Comment by francof2a** [comment, marked as answer]
Hello @Hamuel101,
You're right about that default sizing is `optimal`, and there is a way to define a **template** for new `Fxp` objects:

```python
from fxpmath import Fxp

Fxp.template = Fxp(None, True, 32, 10, op_sizing='same') # enable template

x = Fxp(np.random.random((1,400))) 
y = Fxp(np.random.random((400,128)))

Fxp.template = None # disable template
```

But checking with your example, I found that `config` parameters are set with default values, so `op_sizing` is set with 'optimal' again (I'm going to open an issue with this).

There is another concern is about your example. The function `np.dot` doesn't know how to resolve sizing from `op_sizing` property of `x` or `y` (it could be another issue), you can use some alternatives by now:

```python
z = x.dot(y) # op_sizing of x is used
```
or

```python
import fxpmath as fxp

z = fxp.dot(x, y, sizing='same')
```

Any suggestion is very welcome!

**Reply by SReiter149** [reply]
thanks! That worked

**Comment by SReiter149** [comment]
Apologies for another question. I was wondering if there is a way to have optimal sizing but up to a bit cap, so for example it can have a max n_word of 32 but have any amount of n_bits?

**Reply by francof2a** [reply]
Not problem about questions!
By now there is not a limit for output sizing. It could be added in next revision.

The only limit implemented by now is `n_word_max` in `config`, but it's only used when a new Fxp is created without a specified size.
I don't know details about your needs, but a workaround could be **resize** the Fxp obtained from an operation under a condition:

```python
# ...

N_WORD_MAX = 16

# ...

z.resize(n_word=min(z.n_word, N_WORD_MAX)) # where z is the Fxp object to limit its size
```

I think other way is using `on_value_change` callback previously defined in _template_. Inside that callback you can execute the code above to limit the size. (I have not tested this, but if you are interested, I could do it).

**Reply by SReiter149** [reply]
ok thanks! I will look into this and see what I find.

**Comment by SReiter149** [comment]
all of your suggestions have been very helpful and I am now running into an issue with division. I have been doing as you suggested with fxp.dot and other functions like that. I am coming across a problem where I can't find a fxp.div or fxp.divide or anything like that. So I tried doing `fxp.mul(x,1/y)`but 1/y goes outside of the range I am allowing it so overflows. 

`Fxp.template = Fxp(None, True,8 ,7, op_sizing='same')
x = Fxp(0.635432456768467845673456)
y = Fxp(0.13451234523651234)
z = fxp.mul(x,1/y)
print(z)`

it prints: 0.62786865234375

I can only give it 8 bits for other reasons so can not give integer bits for this purpose. Any suggestions? 

On a side note can I specify rounding in the template like, this doesn't seem to work either: Fxp.template = Fxp(None, True,8 ,7, op_sizing='same',rounding = 'around')

**Reply by francof2a** [reply]
The functions for division are `floordiv` (`//` python operator) and `truediv` (`/` python operator), these functions are from `fxpmath.functions`. Numpy function that are implemented by those functions are `floor_divide` and `true_divide`, correspondingly.

Following examples must be working:

```python
z = x / y
z = np.true_divide(x, y)
z = fxp.truediv(x, y)
```

The `np.divide` implements `np.true_divide`. It's true that `divide` is not mapped (dispatched) by now, but it will be added now you realized it!

About _template_ issue, could you tell me which **fxpmath version** are you using?

```python
import fxpmath as fxp

fxp.__version__
```

**Reply by SReiter149** [reply]
yeah sure, I am on version 0.4.0.

**Reply by francof2a** [reply]
Last version 0.4.2 fixes some template issues. So, update maybe works.
keep me posted, please.

**Reply by SReiter149** [reply]
Yeah, updating it helped. Thanks!

**Comment by SReiter149** [comment]
I have another question. again apologies for your time.

`Fxp.template = Fxp(None, True, 8,7, op_sizing='same',rounding = 'around')
x = np.random.random((4,3,5,7))
y = Fxp(x)
z = np.transpose(y, (3,2,0,1))
print(z.shape)
Fxp.template = None`

Error: TypeError: `out` must be a Fxp object!

Is there a reason you guys raise an error instead of just turning the output into an Fxp object? I feel like you guys could just say 

`if not isinstance(out, Fxp):
out = Fxp(out))`

or else is there a way to transpose and Fxp object?

**Reply by francof2a** [reply]
No problem at all about questions!

By now, the `transpose` method takes `axes` parameters only as _kwarg_, so:

```python
z = np.transpose(y, axes=(3,2,0,1))
z = y.transpose(axes=(3,2,0,1))
```

should work.

If no axes is specified, following must work also:
```python
z = y.T
```

> PS: There is no problem about open a new discussion if the new question is not about original topic.
